\chapter{Discussion and Future Perspectives}
\label{07disc}
\colorbox{yellow}{STLL WIP}

\section{Building accessible and automated tools for MC data analysis}

In this work I have shown CyGNAL's capabilities, describing in detail its design and inner mechanisms, and outlining its usefulness with regards to the analysis of \acrshort{mc} datasets.

The main testament for the usefulness of the tool is the fact that it has become a part of routine \acrshort{mc} analyses in our lab. With its support for plain text to FCS inter-compatibility, one can seamlessly integrate with \acrshort{mc} platforms such as Cytobank. Additionally, as the user only need to run simple Python commands on the terminal to use CyGNAL, it has been readily adopted in day-to-day lab use even by users with no advanced computing experience. 
As I have shown in Chapters \ref{02methods} and \ref{03cytof}, CyGNAL is able to perform a comprehensive analysis of changes occurring across the samples of the often wide \acrshort{mc} experimental systems. Designed with a studying \acrshort{ptm} signalling changes in particular, CyGNAL's computation of \acrshort{emd} and \acrshort{dremi} scores resolves marker intensity and connectivity changes. The easy to use and customisable interactive Shiny-Apps allow for both exploratory and close to publication-grade visualisation of the results. 

Tools are meant to be used, and that publications by colleagues such as Michelozzi et \emph{al.}~\cite{michelozzi_activation_2023} employed CyGNAL is a testament to its relevance.

Originally meant as a simple exercise in curiosity-driven exploration after noticing the correlation between so called PTM and "cell-state" markers, and empowered by the tediousness of manually gating the datasets in our lab, the \acrshort{rf} cell-state classifier has become a convenient tool to automate cell-state labelling of \acrshort{mc} datasets in relation to cell-cycle phases.

Built around a simple \acrfull{rf} architecture, the \acrshort{rf} classifier benefits from the fundamental gate-like logic of both decision trees and the manual cell-state gating process. However, I expect the classifier to suffer from generalisation issues when dealing with external data labelled using different workflows. Furthermore, even if it leverages fuzzy logic to match channel names from the model to the input data, the classifier still relies on matching markers found in both the training data and the data to label. While the markers dedicated to apoptosis and cell-cycle phases generally belong to the less variable portions of \acrshort{mc} panel design, this can still pose an inconvenience when deploying the model. However, I have also shown how other weak points such as low performance for apoptotic class prediction using the 5-marker \acrshort{mc} model, can be effectively addressed by just the addition of an additional apoptotic marker to the panel design. Furthermore, the model seems resilient to cell-type composition and even to broad cell-state changes induced by chemotherapy.

Hearkening back to the link between \acrshort{ptm}s and cell-cycle, the 10-marker \acrshort{mc} model also reveals how certain \acrshort{ptm}s prove more informative when training than \emph{bona fide} cell-cycle markers, and discrepancies between expected state and \acrshort{ptm} correlations from the literature and feature importance rankings have anecdotally been used to validate under-performing antibodies with high unspecific background staining.

Both these tools remain relatively under continuous support, and the idea is to eventually merge both code bases and integrate automated cell-state classification into CyGNAL using pre-built classifier models or allowing for the generation of new models based on specific user-provided labelled data.
CyGNAL could also be augmented by the addition of PHATE~\cite{moon_visualizing_2019} as an alternative dimensionality reduction step, and implementing a third type of Shiny-App to visualise these embeddings and overlay user-selected metadata labels or antibody intensities.

\section{Charting stromal and oncogenic regulation of colonic stem cell polarisation}


scRNA-seq not enough.
% The colonic organoid systems have been thoroughly characterised before at the mass cytometry level and, to better understand these systems, we aim to perform a comparative characterisation of the organoids using scRNA-seq.
Then analysis presented here was paired with mass cytometry. Interregotate regulatino and big scale to functionally understand mecahnisms driving polarisation.

% In summary, through single-cell perturbation analysis of >1,000 organoid cultures, we charted a continuous landscape of cell-intrinsic and -extrinsic regulation of colonic stem cell polarisation. We found that colonic stem cell polarity is regulated by competing YAP and PI3K signalling flux, with stromal TGF-\textbeta\hspace{0.1cm} pushing epithelia towards revCSC and CRC mutations trapping epithelia as proCSC. We conclude that cell-fate plasticity is a hallmark of colonic oncogenesis, and that cells can rapidly traverse the colonic differentiation landscape via combinations of oncogenic and stromal signalling.


        % Single-cell technologies can describe cell-type-specific regulation of differentiation and cell-cell communication \cite{qin_deciphering_2021, fleck_inferring_2022, bues_deterministic_2022}. In this study, we utilised both multiplexed scRNA-seq and high-throughput MC to functionally map how oncogenic mutations and stromal cues co-regulate colonic epithelia across a continuous polarisation landscape. By analysing >1,000 organoid cultures at single-cell resolution, we identify a stepwise cell-fate trajectory spanning from fibroblast-induced revCSC through an equilibrium of balanced differentiation to oncogene-driven proCSC. While scRNA-seq provides in-depth description of colonic epithelial differentiation and proCSC/revCSC polarisation, multiplexed TOB\textit{is} MC allows comprehensive functional interrogation of cell-intrinsic and -extrinsic cues regulating each cell-fate.
        
        % The intestinal stroma comprises a heterogenous population of fibroblasts that regulate the intestinal stem cell niche \cite{gehart_tales_2019}. In the colonic epithelium, CD34\textsuperscript{hi} fibroblasts located at the crypt bottom are a major source of WNT2B, GREM1, and R-Spondin-1, contributing to both homeostatic stem cell maintenance and tissue regeneration following injury \cite{stzepourginski_cd34_2017}. In contrast, CD34\textsuperscript{lo} fibroblasts reside around upper crypts, show lower expression of WNT2B/GREM1 but higher expression of BMPs, thereby providing a permissive environment for epithelial differentiation \cite{ayyaz_singlecell_2019, karpus_colonic_2019}. The fibroblasts used in this study contain both CD34\textsuperscript{hi} and CD34\textsuperscript{lo} cells -- mimicking \textit{in vivo} heterogeneity (Figure \ref{fig:fig1}B). Both CD34\textsuperscript{hi} and CD34\textsuperscript{lo} fibroblast subpopulations showed comparable polarisation of revCSC (Figure \ref{suppfig:figs1}B), suggesting the stromal-epithelial communication in organoid co-cultures may be dominated by TGF-\textbeta1 signalling (Figure \ref{fig:fig6}B). While this study uses healthy colonic fibroblasts to model homeostatic signalling, it is possible cancer associated fibroblasts (CAFs) will communicate differently with epithelial cells, particularly in CRC. Future cell-cell communication studies between CAF sub-types \cite{sahai_framework_2020} and defined epithelial genotypes could uncover exceptions to the signalling models described here and therefore provide novel avenues for therapeutic intervention in CRC.
        
        % Moreover, \textit{shApc} cannot induce revCSC cell-autonomously, indicating revCSC is not immediately downstream of canonical APC/\textbeta-catenin signalling (Figure \ref{suppfig:figs3}A).
        % Collectively, these observations confirmed that organoid cell-fates can be fine-tuned via competing signalling pathways and organoid culture media should be carefully considered when modelling cell-types of interest (Figures \ref{fig:fig3}A, \ref{suppfig:figs4}C-F). 
        
        % proCSC are enriched in CRC organoids and are transcriptionally similar to cells found in human and mouse CRC (Figure \ref{suppfig:figs2}A). However, we demonstrated that proCSC are also present in WT epithelia and highly enriched in WT organoids cultured with WENR ligands. We therefore do not consider proCSC to be cancer stem cells. Rather than establishing an entirely new cancer-specific cell-fate, our study suggests that oncogenic mutations cell-intrinsically polarise cells to an extreme yet pre-existing proCSC state, while simultaneously disrupting cell-extrinsic regulation of plasticity -- trapping cells as proCSC. These results describe cancer as a chronic, unidirectional shift in de-differentiation.
        
        % Although revCSC are most easily accessible in WT epithelia, multiple studies have suggested revCSC also have an important role in CRC \cite{vasquez_dynamic_2022}. revCSC are candidates for early tumour initiating cells \cite{roulis_paracrine_2020} and may confer WNT-inhibitor resistance in CRC \cite{han_lineage_2020}. A recent study in human CRC organoids also demonstrated that cancer cells can escape chemotherapy by adopting a slow-proliferating Mex3a\textsuperscript{+} state driven by a low-EGF and high TGF-\textbeta\hspace{0.1cm}culture environment \cite{alvarez-varela_mex3a_2022}. Our results confirmed that TGF-\textbeta\hspace{0.1cm}can induce revCSC-like cells in CRC organoids, but this process is rare (Figure \ref{suppfig:figs3}C) and requires low PI3K signalling (Figure \ref{fig:fig5}F).
        
                % Moreover, we recently demonstrated that cancer associated fibroblasts (CAFs) can also induce a revCSC-like state in CRC patient-derived organoids (PDOs) that protects CRC cells from chemotherapies including fluorouracil, oxaliplatin, and irinotecan \cite{zapatero_trellis_2022}. In this model, CAF-chemoprotection can also be overcome by inhibiting YAP signalling -- further demonstrating the central role of YAP in revCSC identity. However, CAF-chemoprotection is highly patient-specific, indicating only certain cell-states can be polarised to revCSC in CRC.
                
                % Collectively, our results and others suggest fibroblast-induced revCSCs may represent an important 'drug-tolerant persister' (DTP) state in CRC. Given that targeting cell-plasticity is an emerging area of cancer therapies \cite{burkhardt_mapping_2022}, future studies could target CRC DTP cells by combining YAP inhibitors (to block access to DTP revCSC) with standard chemotherapies (to kill proCSC). 
        
        
        % In this study, we utilised both multiplexed scRNA-seq and high-throughput MC to functionally map how oncogenic mutations and stromal cues co-regulate colonic epithelia across a continuous polarisation landscape.
        
        % This study charts a continuous polarisation trajectory between revCSC and proCSC in colonic epithelia. In the healthy small intestine, revival stem cells have been demonstrated to act as multipotent stem cells that can be mobilised to replenish traditional LGR5\textsuperscript{+} stem cells in response to tissue damage \cite{ayyaz_singlecell_2019}. Small intestinal revival stem cells are found in the homeostatic small intestine \textit{in vivo} \cite{roulis_paracrine_2020,bues_deterministic_2022} and resemble an early 'foetal' stem cell-fate \cite{mustata_identification_2013,yui_yap_2018}. Here we show that in colonic epithelia, revCSC are enriched by fibroblast-derived WNT3A and TGF-\textbeta\hspace{0.1cm}via epithelial YAP, but only in the context of low PI3K and MAPK signalling. Our work and others now collectively suggest that fibroblasts are master regulators of revival stem cells in both the small intestine and colon.

\section{VR Waddington-like landscapes as a data-driven view of scRNA-seq data}

VR land as nbdev project, just like pyKrackhardt


\section{Knowledge Graph}

Coupled with emerging multimodal approaches that can capture at once intra and intercellular communication features, like phospho-seq \cite{blair_phospho-seq_2023} or Jamie's signal-seq. -> perfect tool to solve inter- and intra-cell communications at the single cell level without relying on clusters and in a modality agnostic way. 

